pennylane/qnodes/base.py

```diff
@@ -336,13 +374,49 @@ def _construct(self, args, kwargs):
                 inside of this method, the quantum function returns incorrect values or if
                 both continuous and discrete operations are specified in the same quantum circuit
         """
-        # pylint: disable=attribute-defined-outside-init, too-many-branches
+        # pylint: disable=attribute-defined-outside-init, too-many-branches, too-many-statements
+
+        # flatten the args, replace each argument with a Variable instance carrying a unique index and name
+        full_argspec = inspect.getfullargspec(self.func)
```

Comment:
Could this function call be replaced by adding more functionality to `_get_signature`, which is only called once when the BaseQNode is constructed?

Reply:
The variable names never change (right?), so I think it'd make sense to build the name strings only once, not each time `_construct` is called. I think the same might apply for `arg_vars` and `kwarg_vars` in general, assuming the nested structure of the arguments isn't allowed to vary between qfunc calls?

Reply:
I guess you're correct abouth the `getfullargspec`, that only has to be called once. But `arg_vars` and `kwarg_vars` can only be constructed during `_construct`, because without the user supplying arguments, we don't know their nested structure.

If one would enforce that parameters structure does not change (which is sensible in my opinion), we could cache the `arg_vars` and `kwarg_vars` and only construct them during the first pass.

Reply:
@johannesjmeyer, that reminds me - does the circuit drawer construct code still work when `mutable=False`? QNodes that are immutable cache the result of the contract method and only ever call it once.

Reply:
I see no reason why it should not. As long construct is called once everything is fine? The important thing is that the Variables end up in the circuit_graph and that happens during construction.

Reply:
I implemented a caching logic. This also avoids the flattening step!